### PR TITLE
James dev

### PR DESCRIPTION
**Pull request by zhjwy9343**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [X] Code is well-documented
- [X] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/dotgatconv.py [hunk 1]
```diff
@@ -3,12 +3,13 @@
 from torch import nn
 
 from .... import function as fn
-from ....ops import edge_softmax
+from ..softmax import edge_softmax
```
Comment (yzh119) [inline python/dgl/nn/pytorch/conv/dotgatconv.py:6]: Please import `edge_softmax` from `ops`, `dgl.nn.softmax` is being deprecated.

File: python/dgl/nn/pytorch/glob.py [hunk 2]
```diff
@@ -47,30 +92,76 @@ def forward(self, graph, feat):
...
-        r"""Compute average pooling.
+        r"""
+
+        Compute average pooling.
 
         Parameters
         ----------
-        graph : DGLGraph
-            The graph.
+        graph : DGLGraph or a batch of DGLGraphs
```
Comment (yzh119) [inline python/dgl/nn/pytorch/glob.py:154]: Why mentioning a batch of DGLGraphs? As far as I know it only accepts a single DGLGraph (batched or not batched).

File: python/dgl/nn/pytorch/glob.py [hunk 3]
```diff
@@ -92,10 +227,10 @@ def forward(self, graph, feat):
...
```
Comment (yzh119) [inline python/dgl/nn/pytorch/glob.py:230]: ditto.

File: python/dgl/nn/pytorch/glob.py [hunk 4]
```diff
@@ -111,34 +246,79 @@ def forward(self, graph, feat):
...
-        r"""Compute sort pooling.
+        r"""
+
+        Compute sort pooling.
 
         Parameters
         ----------
-        graph : DGLGraph
-            The graph.
+        graph : DGLGraph or a batch of DGLGraphs
```
Comment (yzh119) [inline python/dgl/nn/pytorch/glob.py:311]: ditto.

File: python/dgl/nn/pytorch/glob.py [hunk 5]
```diff
@@ -163,30 +347,32 @@ class GlobalAttentionPooling(nn.Module):
...
-        r"""Compute global attention pooling.
+        r"""
+
+        Compute global attention pooling.
 
         Parameters
         ----------
-        graph : DGLGraph
-            The graph.
+        graph : DGLGraph or a batch of DGLGraphs
```
Comment (yzh119) [inline python/dgl/nn/pytorch/glob.py:365]: ditto

File: python/dgl/ops/edge_softmax.py [hunk 6]
```diff
@@ -17,83 +19,84 @@ def edge_softmax(graph, logits, eids=ALL, norm_by='dst'):
     called logits in the context of softmax. :math:`\mathcal{N}(i)` is
     the set of nodes that have an edge to :math:`i`.
 
-    By default edge softmax is normalized by destination nodes(i.e. :math:`ij`
-    are incoming edges of `i` in the formula above). We also support edge
-    softmax normalized by source nodes(i.e. :math:`ij` are outgoing edges of
-    `i` in the formula). The previous case correspond to softmax in GAT and
-    Transformer, and the later case correspond to softmax in Capsule network.
+    An example of using edge softmax is in
```
Comment (yzh119) [inline python/dgl/ops/edge_softmax.py:22]: Please do not remove the description of `norm_by`'s behaviour.

File: python/dgl/ops/edge_softmax.py [hunk 7]
```diff
@@ -17,83 +19,84 @@ def edge_softmax(graph, logits, eids=ALL, norm_by='dst'):
...
-    gidx : HeteroGraphIndex
-        The graph to perfor edge softmax on.
+    graph : DGLGraph
+        The graph to perform edge softmax on.
     logits : torch.Tensor
-        The input edge feature
+        The input edge feature.
     eids : torch.Tensor or ALL, optional
-        Edges on which to apply edge softmax. If ALL, apply edge
+        A list of edges on which to apply edge softmax. If ALL, apply edge
```
Comment (yzh119) [inline python/dgl/ops/edge_softmax.py:33]: It's a tensor rather then a list.

File: python/dgl/nn/pytorch/conv/dotgatconv.py [hunk 8]
```diff
@@ -84,8 +95,7 @@ def forward(self, graph, feat):
         graph.edata['sa'] = edge_softmax(graph, graph.edata['a'])
 
         # Step 3. Broadcast softmax value to each edge, and then attention is done
-        graph.apply_edges(lambda edges: {'attn': edges.src['ft'] * \
-                                                 edges.data['sa'].unsqueeze(dim=0).T})
+        graph.apply_edges(fn.u_mul_e('ft', 'sa', 'attn'))
```
Comment (yzh119) [inline python/dgl/nn/pytorch/conv/dotgatconv.py:98]: Please merge this step with the following step by:
```python
graph.update_all(fn.u_mul_e('ft', 'sa', 'attn'), fn.sum('attn', 'agg_u'))
```
to remove the overhead of materializing ft * sa on edges.

File: python/dgl/nn/pytorch/conv/dotgatconv.py [hunk 9]
```diff
@@ -3,12 +3,13 @@
 from torch import nn
 
 from .... import function as fn
-from ....ops import edge_softmax
+from ..softmax import edge_softmax
 from ....utils import expand_as_pair
 
 
 class DotGatConv(nn.Module):
```
Comment (yzh119) [inline python/dgl/nn/pytorch/conv/dotgatconv.py:10]: btw, I was wondering where do we need this module? I think it's exactly the same as Transformer which we have already supported.
